<a href="https://colab.research.google.com/github/Lindal22/Learn-Pandas/blob/master/zerotodatascience_pandas_practice.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Assignment - Pandas Data Analysis Practice

*This assignment is a part of the ["Zero to Data Science Bootcamp by Jovian"](https://zerotodatascience.com)*

In this assignment, you'll get to practice some of the concepts and skills covered this tutorial: https://jovian.ml/aakashns/python-pandas-data-analysis

As you go through this notebook, you will find the symbol **???** in certain places. To complete this assignment, you must replace all the **???** with appropriate values, expressions or statements to ensure that the notebook runs properly end-to-end. 

**Guidelines**

1. Make sure to run all the code cells, otherwise you may get errors like `NameError` for undefined variables.
2. Do not change variable names, delete cells or disturb other existing code. It may cause problems during evaluation.
3. In some cases, you may need to add some code cells or new statements before or after the line of code containing the **???**. 
4. Since you'll be using a temporary online service for code execution, save your work by running `jovian.commit` at regular intervals.
5. Questions marked **(Optional)** will not be considered for evaluation, and can be skipped. They are for your learning.
6. If you are stuck, you can ask for help on the bootcamp Slack group. Post errors, ask for hints and help others, but **please don't share the full solution answer code on Slack** to give others a chance to write the code themselves.
7. There are some tests included with this notebook to help you test your implementation. However, after submission your code will be tested with some hidden test cases. Make sure to test your code exhaustively to cover all edge cases.


**Important Links**

* Submit your work here: https://jovian.ai/learn/zero-to-data-analyst-bootcamp/assignment/pandas-data-analysis-practice/submit
* Review Lesson: https://jovian.ai/learn/zero-to-data-analyst-bootcamp/lesson/3-2-analyzing-tabular-data-with-pandas
* Additional exercises: https://github.com/guipsamora/pandas_exercises


### How to Run the Code and Save Your Work

**Option 1: Running using free online resources (1-click, recommended)**: Click the **Run** button at the top of this page and select **Run on Binder**. You can also select "Run on Colab" or "Run on Kaggle", but you'll need to create an account on [Google Colab](https://colab.research.google.com) or [Kaggle](https://kaggle.com) to use these platforms.


**Option 2: Running on your computer locally**: To run the code on your computer locally, you'll need to set up [Python](https://www.python.org) & [Conda](https://docs.conda.io/projects/conda/en/latest/user-guide/install/), download the notebook and install the required libraries. Click the **Run** button at the top of this page, select the **Run Locally** option, and follow the instructions.

**Saving your work**: You can save a snapshot of the assignment to your [Jovian](https://jovian.ai) profile, so that you can access it later and continue your work. Keep saving your work by running `jovian.commit` from time to time.

In [1]:
# Run the next line to install Pandas
!pip install pandas --upgrade --quiet

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.2/12.2 MB 70.6 MB/s eta 0:00:00


In [2]:
import pandas as pd

In this assignment, we're going to analyze an operate on data from a CSV file. Let's begin by downloading the CSV file.

In [3]:
from urllib.request import urlretrieve

data_url = 'https://gist.githubusercontent.com/aakashns/28b2e504b3350afd9bdb157893f9725c/raw/994b65665757f4f8887db1c85986a897abb23d84/countries.csv'

urlretrieve(data_url, 'countries.csv') # this step is to download the csv file

('countries.csv', <http.client.HTTPMessage at 0x7fe8e2e39430>)

Let's load the data from the CSV file into a Pandas data frame.

In [4]:
countries_df = pd.read_csv('countries.csv')

In [5]:
countries_df.tail(10)

location      continent  population  life_expectancy  \
200  United States Virgin Islands  North America    104423.0            80.58   
201                       Uruguay  South America   3473727.0            77.91   
202                    Uzbekistan           Asia  33469199.0            71.72   
203                       Vatican         Europe       809.0            75.12   
204                     Venezuela  South America  28435943.0            72.06   
205                       Vietnam           Asia  97338583.0            75.40   
206                Western Sahara         Africa    597330.0            70.26   
207                         Yemen           Asia  29825968.0            66.12   
208                        Zambia         Africa  18383956.0            63.89   
209                      Zimbabwe         Africa  14862927.0            61.49   

     hospital_beds_per_thousand  gdp_per_capita  
200                         NaN             NaN  
201                         2.8       20551.409  
202                         4.0        6253.104  
203                         NaN             NaN  
204                         0.8       16745.022  
205                         2.6        6171.884  
206                         NaN             NaN  
207                         0.7        1479.147  
208                         2.0        3689.251  
209                         1.7        1899.775

**Q1: How many countries does the dataframe contain?**

Hint: Use the `.shape` method.

In [6]:
countries_df.shape 

(210, 6)

In [7]:
num_countries = countries_df.shape[0] # shape method returns a turple, we only need the first item.

In [8]:
print('There are {} countries in the dataset'.format(num_countries))

There are 210 countries in the dataset


**Q2: Retrieve a list of continents from the dataframe?**

*Hint: Use the `.unique` method of a series.*

In [9]:
countries_df.columns

Index(['location', 'continent', 'population', 'life_expectancy',
       'hospital_beds_per_thousand', 'gdp_per_capita'],
      dtype='object')

In [10]:
continents = countries_df['continent'].unique()

In [11]:
continents

array(['Asia', 'Europe', 'Africa', 'North America', 'South America',
       'Oceania'], dtype=object)

**Q3: What is the total population of all the countries listed in this dataset?**

In [12]:
total_population = countries_df['population'].sum() # Attention ,this sum() function is pandas 
                                                    # function working on dataframe's columns 
                                                    #  -- series

In [13]:
total_population

7757980095.0

In [14]:
print('The total population is {}.'.format(int(total_population)))

The total population is 7757980095.


**Q: (Optional) What is the overall life expectancy across in the world?**

*Hint: You'll need to take a weighted average of life expectancy using populations as weights.*

In [15]:
import numpy as np

In [16]:
population = countries_df.population
life_expectancy = countries_df.life_expectancy

In [17]:
np.sum(np.multiply(population,life_expectancy))/np.sum(countries_df.population)

72.72165193409664

In [18]:
avg_life_expectancy = np.sum(np.multiply(population,life_expectancy))/total_population

In [19]:
print("\nThe average life expectancy for the whole world is {:.2f}.\n ".format(avg_life_expectancy))


The average life expectancy for the whole world is 72.72.
 


**Q4: Create a dataframe containing 10 countries with the highest population.**

*Hint: Chain the `sort_values` and `head` methods.*

In [20]:
most_popul_df = countries_df.sort_values(by=['population'], ascending=False).head(10)


In [21]:
most_populous_df = countries_df.sort_values(by=['population'], ascending=False).head(10)

In [22]:
most_popul_df

location      continent    population  life_expectancy  \
41           China           Asia  1.439324e+09            76.91   
90           India           Asia  1.380004e+09            69.66   
199  United States  North America  3.310026e+08            78.86   
91       Indonesia           Asia  2.735236e+08            71.72   
145       Pakistan           Asia  2.208923e+08            67.27   
27          Brazil  South America  2.125594e+08            75.88   
141        Nigeria         Africa  2.061396e+08            54.69   
15      Bangladesh           Asia  1.646894e+08            72.59   
157         Russia         Europe  1.459345e+08            72.58   
125         Mexico  North America  1.289328e+08            75.05   

     hospital_beds_per_thousand  gdp_per_capita  
41                         4.34       15308.712  
90                         0.53        6426.674  
199                        2.77       54225.446  
91                         1.04       11188.744  
145                        0.60        5034.708  
27                         2.20       14103.452  
141                         NaN        5338.454  
15                         0.80        3523.984  
157                        8.05       24765.954  
125                        1.38       17336.469

**Q5: Add a new column in `countries_df` to record the overall GDP per country (product of population & per capita GDP).**



In [23]:
np.multiply(countries_df.population, countries_df.gdp_per_capita)

0      7.022622e+10
1      3.396791e+10
2      6.101364e+11
3               NaN
4      1.912651e+11
           ...     
205    6.007624e+11
206             NaN
207    4.411699e+10
208    6.782303e+10
209    2.823622e+10
Length: 210, dtype: float64

In [24]:
countries_df['gdp'] = np.multiply(countries_df.population, countries_df.gdp_per_capita)

In [25]:
countries_df

location continent  population  life_expectancy  \
0       Afghanistan      Asia  38928341.0            64.83   
1           Albania    Europe   2877800.0            78.57   
2           Algeria    Africa  43851043.0            76.88   
3           Andorra    Europe     77265.0            83.73   
4            Angola    Africa  32866268.0            61.15   
..              ...       ...         ...              ...   
205         Vietnam      Asia  97338583.0            75.40   
206  Western Sahara    Africa    597330.0            70.26   
207           Yemen      Asia  29825968.0            66.12   
208          Zambia    Africa  18383956.0            63.89   
209        Zimbabwe    Africa  14862927.0            61.49   

     hospital_beds_per_thousand  gdp_per_capita           gdp  
0                          0.50        1803.987  7.022622e+10  
1                          2.89       11803.431  3.396791e+10  
2                          1.90       13913.839  6.101364e+11  
3                           NaN             NaN           NaN  
4                           NaN        5819.495  1.912651e+11  
..                          ...             ...           ...  
205                        2.60        6171.884  6.007624e+11  
206                         NaN             NaN           NaN  
207                        0.70        1479.147  4.411699e+10  
208                        2.00        3689.251  6.782303e+10  
209                        1.70        1899.775  2.823622e+10  

[210 rows x 7 columns]

**Q: (Optional) Create a dataframe containing 10 countries with the lowest GDP per capita, among the counties with population greater than 100 million.**

In [26]:
lower_gdp_df = countries_df[population > 100000000].sort_values(by=['gdp_per_capita'],ascending=True).head(10)

In [27]:
lower_gdp_df

location      continent    population  life_expectancy  \
63      Ethiopia         Africa  1.149636e+08            66.60   
15    Bangladesh           Asia  1.646894e+08            72.59   
145     Pakistan           Asia  2.208923e+08            67.27   
141      Nigeria         Africa  2.061396e+08            54.69   
90         India           Asia  1.380004e+09            69.66   
151  Philippines           Asia  1.095811e+08            71.23   
58         Egypt         Africa  1.023344e+08            71.99   
91     Indonesia           Asia  2.735236e+08            71.72   
27        Brazil  South America  2.125594e+08            75.88   
41         China           Asia  1.439324e+09            76.91   

     hospital_beds_per_thousand  gdp_per_capita           gdp  
63                         0.30        1729.927  1.988786e+11  
15                         0.80        3523.984  5.803628e+11  
145                        0.60        5034.708  1.112128e+12  
141                         NaN        5338.454  1.100467e+12  
90                         0.53        6426.674  8.868838e+12  
151                        1.00        7599.188  8.327273e+11  
58                         1.60       10550.206  1.079649e+12  
91                         1.04       11188.744  3.060386e+12  
27                         2.20       14103.452  2.997821e+12  
41                         4.34       15308.712  2.203419e+13

**Q6: Create a data frame that counts the number countries in each continent?**

*Hint: Use `groupby`, select the `location` column and aggregate using `count`.*

In [28]:
count_countries_df = countries_df.groupby('continent')['location'].count()
count_countries_df

continent
Africa           55
Asia             47
Europe           51
North America    36
Oceania           8
South America    13
Name: location, dtype: int64

In [29]:
country_counts_df = countries_df.groupby('continent')['location'].count()

In [30]:
country_counts_df

continent
Africa           55
Asia             47
Europe           51
North America    36
Oceania           8
South America    13
Name: location, dtype: int64

**Q7: Create a data frame showing the total population of each continent.**

*Hint: Use `groupby`, select the population column and aggregate using `sum`.*

In [31]:
cont_popu_df = countries_df.groupby('continent')['population'].sum()
cont_popu_df

continent
Africa           1.339424e+09
Asia             4.607388e+09
Europe           7.485062e+08
North America    5.912425e+08
Oceania          4.095832e+07
South America    4.304611e+08
Name: population, dtype: float64

In [32]:
continent_populations_df = countries_df.groupby('continent')['population'].sum()

In [33]:
continent_populations_df

continent
Africa           1.339424e+09
Asia             4.607388e+09
Europe           7.485062e+08
North America    5.912425e+08
Oceania          4.095832e+07
South America    4.304611e+08
Name: population, dtype: float64

Let's download another CSV file containing overall Covid-19 stats for various countires, and read the data into another Pandas data frame.

In [34]:
covid_data_url = 'https://gist.githubusercontent.com/aakashns/b2a968a6cfd9fbbb0ff3d6bd0f26262b/raw/b115ed1dfa17f10fc88bf966236cd4d9032f1df8/covid-countries-data.csv'

urlretrieve(covid_data_url, 'covid-countries-data.csv')

('covid-countries-data.csv', <http.client.HTTPMessage at 0x7fe8e2d418e0>)

In [35]:
covid_data_df = pd.read_csv('covid-countries-data.csv')

In [36]:
covid_data_df.sample(11)

location  total_cases  total_deaths  total_tests
191                       Togo       1434.0          30.0      54709.0
94                        Iraq     242284.0        7201.0    1624331.0
171  Sint Maarten (Dutch part)        482.0          19.0          NaN
88                     Hungary       6622.0         619.0     437531.0
125                  Mauritius        356.0          10.0          NaN
143   Northern Mariana Islands         57.0           2.0          NaN
8                      Armenia      44075.0         884.0          NaN
185                      Syria       2898.0         120.0          NaN
113                      Libya      15156.0         250.0          NaN
118                 Madagascar      15023.0         196.0          NaN
156                      Qatar     119206.0         199.0     634745.0

In [37]:
covid_data_df[['location','total_tests']][covid_data_df.total_tests == 'nan']

Empty DataFrame
Columns: [location, total_tests]
Index: []

In [38]:
covid_data_df.sample(11)

location  total_cases  total_deaths  total_tests
23                    Bolivia     117928.0        5203.0     238965.0
182                 Swaziland       4668.0          94.0          NaN
143  Northern Mariana Islands         57.0           2.0          NaN
76                     Greece      10757.0         273.0     966346.0
26                   Botswana       1724.0           6.0          NaN
62                    Estonia       2415.0          64.0     151940.0
144                    Norway      10871.0         264.0     711225.0
175              South Africa     630595.0       14389.0    3705408.0
79                       Guam       1447.0          13.0          NaN
185                     Syria       2898.0         120.0          NaN
119                    Malawi       5579.0         175.0          NaN

In [39]:
covid_data_df.set_index('location',inplace=True)

In [40]:
covid_data_df

total_cases  total_deaths  total_tests
location                                              
Afghanistan         38243.0        1409.0          NaN
Albania              9728.0         296.0          NaN
Algeria             45158.0        1525.0          NaN
Andorra              1199.0          53.0          NaN
Angola               2729.0         109.0          NaN
...                     ...           ...          ...
Western Sahara        766.0           1.0          NaN
World            26059065.0      863535.0          NaN
Yemen                1976.0         571.0          NaN
Zambia              12415.0         292.0          NaN
Zimbabwe             6638.0         206.0      97272.0

[212 rows x 3 columns]

In [47]:
covid_data_df.reset_index()

location  total_cases  total_deaths  total_tests
0       Afghanistan      38243.0        1409.0          NaN
1           Albania       9728.0         296.0          NaN
2           Algeria      45158.0        1525.0          NaN
3           Andorra       1199.0          53.0          NaN
4            Angola       2729.0         109.0          NaN
..              ...          ...           ...          ...
207  Western Sahara        766.0           1.0          NaN
208           World   26059065.0      863535.0          NaN
209           Yemen       1976.0         571.0          NaN
210          Zambia      12415.0         292.0          NaN
211        Zimbabwe       6638.0         206.0      97272.0

[212 rows x 4 columns]

In [49]:
covid_data_df

total_cases  total_deaths  total_tests
location                                              
Afghanistan         38243.0        1409.0          NaN
Albania              9728.0         296.0          NaN
Algeria             45158.0        1525.0          NaN
Andorra              1199.0          53.0          NaN
Angola               2729.0         109.0          NaN
...                     ...           ...          ...
Western Sahara        766.0           1.0          NaN
World            26059065.0      863535.0          NaN
Yemen                1976.0         571.0          NaN
Zambia              12415.0         292.0          NaN
Zimbabwe             6638.0         206.0      97272.0

[212 rows x 3 columns]

In [43]:
covid_data_df.total_tests

location
Afghanistan           NaN
Albania               NaN
Algeria               NaN
Andorra               NaN
Angola                NaN
                   ...   
Western Sahara        NaN
World                 NaN
Yemen                 NaN
Zambia                NaN
Zimbabwe          97272.0
Name: total_tests, Length: 212, dtype: float64

In [44]:
covid_data_df[covid_data_df.total_tests.isna()].count()


total_cases     122
total_deaths    122
total_tests       0
dtype: int64

**Q8: Count the number of countries for which the `total_tests` data is missing.**

*Hint: Use the `.isna` method.*

In [52]:
total_tests_missing = covid_data_df[covid_data_df.total_tests.isna()].count()
total_tests_missing

total_cases     122
total_deaths    122
total_tests       0
dtype: int64

In [54]:
print("The data for total tests is missing for {} countries.".format(total_tests_missing))

The data for total tests is missing for total_cases     122
total_deaths    122
total_tests       0
dtype: int64 countries.


Let's merge the two data frames, and compute some more metrics.

**Q9: Merge `countries_df` with `covid_data_df` on the `location` column.**

*Hint: Use the `.merge` method on `countries_df`.

In [55]:
countries_df

location continent  population  life_expectancy  \
0       Afghanistan      Asia  38928341.0            64.83   
1           Albania    Europe   2877800.0            78.57   
2           Algeria    Africa  43851043.0            76.88   
3           Andorra    Europe     77265.0            83.73   
4            Angola    Africa  32866268.0            61.15   
..              ...       ...         ...              ...   
205         Vietnam      Asia  97338583.0            75.40   
206  Western Sahara    Africa    597330.0            70.26   
207           Yemen      Asia  29825968.0            66.12   
208          Zambia    Africa  18383956.0            63.89   
209        Zimbabwe    Africa  14862927.0            61.49   

     hospital_beds_per_thousand  gdp_per_capita           gdp  
0                          0.50        1803.987  7.022622e+10  
1                          2.89       11803.431  3.396791e+10  
2                          1.90       13913.839  6.101364e+11  
3                           NaN             NaN           NaN  
4                           NaN        5819.495  1.912651e+11  
..                          ...             ...           ...  
205                        2.60        6171.884  6.007624e+11  
206                         NaN             NaN           NaN  
207                        0.70        1479.147  4.411699e+10  
208                        2.00        3689.251  6.782303e+10  
209                        1.70        1899.775  2.823622e+10  

[210 rows x 7 columns]

In [56]:
covid_data_df

total_cases  total_deaths  total_tests
location                                              
Afghanistan         38243.0        1409.0          NaN
Albania              9728.0         296.0          NaN
Algeria             45158.0        1525.0          NaN
Andorra              1199.0          53.0          NaN
Angola               2729.0         109.0          NaN
...                     ...           ...          ...
Western Sahara        766.0           1.0          NaN
World            26059065.0      863535.0          NaN
Yemen                1976.0         571.0          NaN
Zambia              12415.0         292.0          NaN
Zimbabwe             6638.0         206.0      97272.0

[212 rows x 3 columns]

In [57]:
combined_df = covid_data_df.merge(countries_df, on="location")

In [58]:
combined_df

location  total_cases  total_deaths  total_tests continent  \
0       Afghanistan      38243.0        1409.0          NaN      Asia   
1           Albania       9728.0         296.0          NaN    Europe   
2           Algeria      45158.0        1525.0          NaN    Africa   
3           Andorra       1199.0          53.0          NaN    Europe   
4            Angola       2729.0         109.0          NaN    Africa   
..              ...          ...           ...          ...       ...   
205         Vietnam       1046.0          35.0     261004.0      Asia   
206  Western Sahara        766.0           1.0          NaN    Africa   
207           Yemen       1976.0         571.0          NaN      Asia   
208          Zambia      12415.0         292.0          NaN    Africa   
209        Zimbabwe       6638.0         206.0      97272.0    Africa   

     population  life_expectancy  hospital_beds_per_thousand  gdp_per_capita  \
0    38928341.0            64.83                        0.50        1803.987   
1     2877800.0            78.57                        2.89       11803.431   
2    43851043.0            76.88                        1.90       13913.839   
3       77265.0            83.73                         NaN             NaN   
4    32866268.0            61.15                         NaN        5819.495   
..          ...              ...                         ...             ...   
205  97338583.0            75.40                        2.60        6171.884   
206    597330.0            70.26                         NaN             NaN   
207  29825968.0            66.12                        0.70        1479.147   
208  18383956.0            63.89                        2.00        3689.251   
209  14862927.0            61.49                        1.70        1899.775   

              gdp  
0    7.022622e+10  
1    3.396791e+10  
2    6.101364e+11  
3             NaN  
4    1.912651e+11  
..            ...  
205  6.007624e+11  
206           NaN  
207  4.411699e+10  
208  6.782303e+10  
209  2.823622e+10  

[210 rows x 10 columns]

**Q10: Add columns `tests_per_million`, `cases_per_million` and `deaths_per_million` into `combined_df`.**

In [60]:
combined_df['tests_per_million'] = combined_df['total_tests'] * 1e6 /combined_df['population']

In [61]:
combined_df['cases_per_million'] = combined_df['total_cases'] *1e6 / combined_df['population']

In [62]:
combined_df['deaths_per_million'] = combined_df['total_deaths'] * 1e6 / combined_df['population']

In [63]:
combined_df

location  total_cases  total_deaths  total_tests continent  \
0       Afghanistan      38243.0        1409.0          NaN      Asia   
1           Albania       9728.0         296.0          NaN    Europe   
2           Algeria      45158.0        1525.0          NaN    Africa   
3           Andorra       1199.0          53.0          NaN    Europe   
4            Angola       2729.0         109.0          NaN    Africa   
..              ...          ...           ...          ...       ...   
205         Vietnam       1046.0          35.0     261004.0      Asia   
206  Western Sahara        766.0           1.0          NaN    Africa   
207           Yemen       1976.0         571.0          NaN      Asia   
208          Zambia      12415.0         292.0          NaN    Africa   
209        Zimbabwe       6638.0         206.0      97272.0    Africa   

     population  life_expectancy  hospital_beds_per_thousand  gdp_per_capita  \
0    38928341.0            64.83                        0.50        1803.987   
1     2877800.0            78.57                        2.89       11803.431   
2    43851043.0            76.88                        1.90       13913.839   
3       77265.0            83.73                         NaN             NaN   
4    32866268.0            61.15                         NaN        5819.495   
..          ...              ...                         ...             ...   
205  97338583.0            75.40                        2.60        6171.884   
206    597330.0            70.26                         NaN             NaN   
207  29825968.0            66.12                        0.70        1479.147   
208  18383956.0            63.89                        2.00        3689.251   
209  14862927.0            61.49                        1.70        1899.775   

              gdp  tests_per_million  cases_per_million  deaths_per_million  
0    7.022622e+10                NaN         982.394806           36.194710  
1    3.396791e+10                NaN        3380.359997          102.856349  
2    6.101364e+11                NaN        1029.804468           34.776824  
3             NaN                NaN       15518.022390          685.950948  
4    1.912651e+11                NaN          83.033462            3.316470  
..            ...                ...                ...                 ...  
205  6.007624e+11        2681.403324          10.745996            0.359570  
206           NaN                NaN        1282.373228            1.674116  
207  4.411699e+10                NaN          66.250993           19.144391  
208  6.782303e+10                NaN         675.317108           15.883415  
209  2.823622e+10        6544.605918         446.614587           13.859989  

[210 rows x 13 columns]

**Q11: Create a dataframe with 10 countires that have highest number of tests per million people.**

In [64]:
combined_df.sort_values(by=['tests_per_million'], ascending=False).head(10)

location  total_cases  total_deaths  total_tests  \
197  United Arab Emirates      71540.0         387.0    7177430.0   
14                Bahrain      52440.0         190.0    1118837.0   
115            Luxembourg       7928.0         124.0     385820.0   
122                 Malta       1931.0          13.0     188539.0   
53                Denmark      17195.0         626.0    2447911.0   
96                 Israel     122539.0         969.0    2353984.0   
89                Iceland       2121.0          10.0      88829.0   
157                Russia    1005000.0       17414.0   37176827.0   
199         United States    6114406.0      185744.0   83898416.0   
10              Australia      25923.0         663.0    6255797.0   

         continent   population  life_expectancy  hospital_beds_per_thousand  \
197           Asia    9890400.0            77.97                       1.200   
14            Asia    1701583.0            77.29                       2.000   
115         Europe     625976.0            82.25                       4.510   
122         Europe     441539.0            82.53                       4.485   
53          Europe    5792203.0            80.90                       2.500   
96            Asia    8655541.0            82.97                       2.990   
89          Europe     341250.0            82.99                       2.910   
157         Europe  145934460.0            72.58                       8.050   
199  North America  331002647.0            78.86                       2.770   
10         Oceania   25499881.0            83.44                       3.840   

     gdp_per_capita           gdp  tests_per_million  cases_per_million  \
197       67293.483  6.655595e+11      725696.635121        7233.276713   
14        43290.705  7.366273e+10      657527.137965       30818.361490   
115       94277.965  5.901574e+10      616349.508607       12665.022301   
122       36513.323  1.612206e+10      427004.183096        4373.339614   
53        46682.515  2.703946e+11      422621.755488        2968.645954   
96        33132.320  2.867782e+11      271962.665303       14157.289533   
89        46482.958  1.586231e+10      260304.761905        6215.384615   
157       24765.954  3.614206e+12      254750.159763        6886.653091   
199       54225.446  1.794877e+13      253467.507769       18472.377957   
10        44648.710  1.138537e+12      245326.517406        1016.592979   

     deaths_per_million  
197           39.128852  
14           111.660730  
115          198.090662  
122           29.442473  
53           108.076323  
96           111.951408  
89            29.304029  
157          119.327539  
199          561.155633  
10            26.000121

**Q12: Create a dataframe with 10 countires that have highest number of positive cases per million people.**

In [65]:
highest_cases_df = combined_df.sort_values(by=['tests_per_million'], ascending=False).head(10)

In [66]:
highest_cases_df

location  total_cases  total_deaths  total_tests  \
197  United Arab Emirates      71540.0         387.0    7177430.0   
14                Bahrain      52440.0         190.0    1118837.0   
115            Luxembourg       7928.0         124.0     385820.0   
122                 Malta       1931.0          13.0     188539.0   
53                Denmark      17195.0         626.0    2447911.0   
96                 Israel     122539.0         969.0    2353984.0   
89                Iceland       2121.0          10.0      88829.0   
157                Russia    1005000.0       17414.0   37176827.0   
199         United States    6114406.0      185744.0   83898416.0   
10              Australia      25923.0         663.0    6255797.0   

         continent   population  life_expectancy  hospital_beds_per_thousand  \
197           Asia    9890400.0            77.97                       1.200   
14            Asia    1701583.0            77.29                       2.000   
115         Europe     625976.0            82.25                       4.510   
122         Europe     441539.0            82.53                       4.485   
53          Europe    5792203.0            80.90                       2.500   
96            Asia    8655541.0            82.97                       2.990   
89          Europe     341250.0            82.99                       2.910   
157         Europe  145934460.0            72.58                       8.050   
199  North America  331002647.0            78.86                       2.770   
10         Oceania   25499881.0            83.44                       3.840   

     gdp_per_capita           gdp  tests_per_million  cases_per_million  \
197       67293.483  6.655595e+11      725696.635121        7233.276713   
14        43290.705  7.366273e+10      657527.137965       30818.361490   
115       94277.965  5.901574e+10      616349.508607       12665.022301   
122       36513.323  1.612206e+10      427004.183096        4373.339614   
53        46682.515  2.703946e+11      422621.755488        2968.645954   
96        33132.320  2.867782e+11      271962.665303       14157.289533   
89        46482.958  1.586231e+10      260304.761905        6215.384615   
157       24765.954  3.614206e+12      254750.159763        6886.653091   
199       54225.446  1.794877e+13      253467.507769       18472.377957   
10        44648.710  1.138537e+12      245326.517406        1016.592979   

     deaths_per_million  
197           39.128852  
14           111.660730  
115          198.090662  
122           29.442473  
53           108.076323  
96           111.951408  
89            29.304029  
157          119.327539  
199          561.155633  
10            26.000121

**Q13: Create a dataframe with 10 countires that have highest number of deaths cases per million people?**

In [67]:
combined_df.sort_values(by=['deaths_per_million'],ascending=False).head(10)

location  total_cases  total_deaths  total_tests      continent  \
162      San Marino        735.0          42.0          NaN         Europe   
150            Peru     663437.0       29259.0     584232.0  South America   
18          Belgium      85817.0        9898.0    2281853.0         Europe   
3           Andorra       1199.0          53.0          NaN         Europe   
177           Spain     479554.0       29194.0    6416533.0         Europe   
198  United Kingdom     338676.0       41514.0   13447568.0         Europe   
40            Chile     414739.0       11344.0    2458762.0  South America   
97            Italy     271515.0       35497.0    5214766.0         Europe   
27           Brazil    3997865.0      123780.0    4797948.0  South America   
182          Sweden      84532.0        5820.0          NaN         Europe   

      population  life_expectancy  hospital_beds_per_thousand  gdp_per_capita  \
162      33938.0            84.97                        3.80       56861.470   
150   32971846.0            76.74                        1.60       12236.706   
18    11589616.0            81.63                        5.64       42658.576   
3        77265.0            83.73                         NaN             NaN   
177   46754783.0            83.56                        2.97       34272.360   
198   67886004.0            81.32                        2.54       39753.244   
40    19116209.0            80.18                        2.11       22767.037   
97    60461828.0            83.51                        3.18       35220.084   
27   212559409.0            75.88                        2.20       14103.452   
182   10099270.0            82.80                        2.22       46949.283   

              gdp  tests_per_million  cases_per_million  deaths_per_million  
162  1.929765e+09                NaN       21657.139490         1237.550828  
150  4.034668e+11       17719.117092       20121.318048          887.393445  
18   4.943965e+11      196887.713967        7404.645676          854.040375  
3             NaN                NaN       15518.022390          685.950948  
177  1.602397e+12      137238.001939       10256.790198          624.406705  
198  2.698689e+12      198090.434075        4988.892850          611.525168  
40   4.352194e+11      128621.841287       21695.671982          593.423100  
97   2.129471e+12       86248.897403        4490.684602          587.097697  
27   2.997821e+12       22572.268255       18808.224105          582.331314  
182  4.741535e+11                NaN        8370.109919          576.279276

In [68]:
highest_deaths_df = combined_df.sort_values(by=['deaths_per_million'],ascending=False).head(10)

In [69]:
highest_deaths_df

location  total_cases  total_deaths  total_tests      continent  \
162      San Marino        735.0          42.0          NaN         Europe   
150            Peru     663437.0       29259.0     584232.0  South America   
18          Belgium      85817.0        9898.0    2281853.0         Europe   
3           Andorra       1199.0          53.0          NaN         Europe   
177           Spain     479554.0       29194.0    6416533.0         Europe   
198  United Kingdom     338676.0       41514.0   13447568.0         Europe   
40            Chile     414739.0       11344.0    2458762.0  South America   
97            Italy     271515.0       35497.0    5214766.0         Europe   
27           Brazil    3997865.0      123780.0    4797948.0  South America   
182          Sweden      84532.0        5820.0          NaN         Europe   

      population  life_expectancy  hospital_beds_per_thousand  gdp_per_capita  \
162      33938.0            84.97                        3.80       56861.470   
150   32971846.0            76.74                        1.60       12236.706   
18    11589616.0            81.63                        5.64       42658.576   
3        77265.0            83.73                         NaN             NaN   
177   46754783.0            83.56                        2.97       34272.360   
198   67886004.0            81.32                        2.54       39753.244   
40    19116209.0            80.18                        2.11       22767.037   
97    60461828.0            83.51                        3.18       35220.084   
27   212559409.0            75.88                        2.20       14103.452   
182   10099270.0            82.80                        2.22       46949.283   

              gdp  tests_per_million  cases_per_million  deaths_per_million  
162  1.929765e+09                NaN       21657.139490         1237.550828  
150  4.034668e+11       17719.117092       20121.318048          887.393445  
18   4.943965e+11      196887.713967        7404.645676          854.040375  
3             NaN                NaN       15518.022390          685.950948  
177  1.602397e+12      137238.001939       10256.790198          624.406705  
198  2.698689e+12      198090.434075        4988.892850          611.525168  
40   4.352194e+11      128621.841287       21695.671982          593.423100  
97   2.129471e+12       86248.897403        4490.684602          587.097697  
27   2.997821e+12       22572.268255       18808.224105          582.331314  
182  4.741535e+11                NaN        8370.109919          576.279276

**(Optional) Q: Count number of countries that feature in both the lists of "highest number of tests per million" and "highest number of cases per million".**

In [70]:
countries_highest_deaths = combined_df.sort_values(by=['deaths_per_million'],ascending=False).head(10)['location']
countries_highest_deaths

162        San Marino
150              Peru
18            Belgium
3             Andorra
177             Spain
198    United Kingdom
40              Chile
97              Italy
27             Brazil
182            Sweden
Name: location, dtype: object

In [71]:
countries_highest_tests = combined_df.sort_values(by=['tests_per_million'],ascending=False).head(10)['location']
countries_highest_tests

197    United Arab Emirates
14                  Bahrain
115              Luxembourg
122                   Malta
53                  Denmark
96                   Israel
89                  Iceland
157                  Russia
199           United States
10                Australia
Name: location, dtype: object

In [72]:
countries_highest_cases = combined_df.sort_values(by=['cases_per_million'],ascending=False).head(10)['location']
countries_highest_cases

155            Qatar
14           Bahrain
147           Panama
40             Chile
162       San Marino
9              Aruba
105           Kuwait
150             Peru
27            Brazil
199    United States
Name: location, dtype: object

In [73]:
def no_countries(countries1, countries2):
  results = []
  for country in countries1:
    for c in countries2:
      if country == c:
        results.append(country)
  return len(results)

In [74]:
number_countries = no_countries(countries_highest_tests, countries_highest_cases)

In [75]:
number_countries

2

**(Optional) Q: Count number of countries that feature in both the lists "20 countries with lowest GDP per capita" and "20 countries with the lowest number of hospital beds per thousand population". Only consider countries with a population higher than 10 million while creating the list.**

In [76]:
countr_low_gdp = countries_df[countries_df.population > 10000000].sort_values(by=['gdp_per_capita'],ascending=True).head(20)['location']
countr_low_gdp

32                          Burundi
52     Democratic Republic of Congo
140                           Niger
118                          Malawi
132                      Mozambique
117                      Madagascar
207                           Yemen
176                     South Sudan
85                            Haiti
195                          Uganda
31                     Burkina Faso
63                         Ethiopia
39                             Chad
0                       Afghanistan
158                          Rwanda
209                        Zimbabwe
82                           Guinea
121                            Mali
20                            Benin
135                           Nepal
Name: location, dtype: object

In [77]:
countr_low_beds = countries_df[countries_df.population > 10000000].sort_values(by=['hospital_beds_per_thousand'],ascending=True).head(20)['location']
countr_low_beds 

121            Mali
117      Madagascar
140           Niger
135           Nepal
82           Guinea
63         Ethiopia
31     Burkina Faso
195          Uganda
0       Afghanistan
20            Benin
90            India
80        Guatemala
145        Pakistan
85            Haiti
207           Yemen
187        Tanzania
132      Mozambique
32          Burundi
33         Cambodia
204       Venezuela
Name: location, dtype: object

In [78]:
no_countries(countr_low_gdp, countr_low_beds)

14

## LeetCode Problems

### [Problem 482](https://leetcode.com/problems/license-key-formatting/) - License Key Formatting

You are given a license key represented as a string s that consists of only alphanumeric characters and dashes. The string is separated into n + 1 groups by n dashes. You are also given an integer k.

We want to reformat the string s such that each group contains exactly k characters, except for the first group, which could be shorter than k but still must contain at least one character. Furthermore, there must be a dash inserted between two groups, and you should convert all lowercase letters to uppercase.

Return the reformatted license key.

**Example 1:**
```
Input: s = "5F3Z-2e-9-w", k = 4

Output: "5F3Z-2E9W"

Explanation:  The string s has been split into two parts, each part has 4 characters.
Note that the two extra dashes are not needed and can be removed.
```







**Example 2:**
```
Input: s = "2-5g-3-J", k = 2

Output: "2-5G-3J"

Explanation:  The string s has been split into three parts, each part has 2 characters 
except the first part as it could be shorter as mentioned above.
```





**Constraints:**
```
1 <= s.length <= 105
s consists of English letters, digits, and dashes '-'.
1 <= k <= 104
```

In [ ]:
s = '2-5g-3-J'

In [ ]:
s[:-1]

In [ ]:
str = []
for i in reversed(range(len(s))):
  str += s[i]
str

In [ ]:

list = []
for i in reversed(range(8)):
  list.append(i)
list

In [ ]:
def transform_str(s,k):
    str =[]
    count = 0
    for i in reversed(range(len(s))): # reversed the whole string
      if s[i] != '-': # skip the '-' character
        str.append(s[i].upper()) # upper case the letter and append it to the new string
        count += 1
        if count == k: # if there are k number of elements, stop
          count = 0 # start the next group
          str.append('-') # this group finished, add '-' to the end
    
    if(len(s) > 0 and str[-1] == '-'): # the last group last character should be '-'
       str = str[:-1]   # remove the last character
    str = str[::-1]   # reverse the whole new string
    return str

In [ ]:
def transform_string(s,k): # s is the inpit string, k is the group size
  """transform the input string (only letters, digits, and '-') into groups of string
  get rid of '-'s,group size is k, there are '-' between groups, the first group may be k, or less than k"""
  count = 0 # need a counter
  results = [] # the result string will be returned.
  # starting from the last character, and in reversed order
  for i in reversed(range(len(s))):
    # check if the character is '-', if not, uppercase it and put into the result string, add 1 to count
    if s[i] != '-':
      print("s[i] ",s[i])
      results.append(s[i].upper())
      count = count + 1
      print("count",count)
      # check if count is equal to group size k, if yes, count reset to zero, the result string add '-'
      if count == k:
        count = 0
        results.append('-')

  if (len(s) > 0 and results[-1] == '-'):
    results = results[:-1]
  results = results[::-1] 
  return results


In [ ]:
print(transform_string('a2-5g-3-J',2))

In [ ]:
def licenseKeyFormatting(s, k):
  """transform the input string (only letters, digits, and '-') into groups of string
  get rid of '-'s,group size is k, there are '-' between groups, the first group may be k, or less than k"""
  count = 0 # need a counter
  results = [] # the result string will be returned.
  # starting from the last character, and in reversed order
  for i in reversed(range(len(s))):
    # check if the character is '-', if not, uppercase it and put into the result string, add 1 to count
    if s[i] != '-':

      results.append(s[i].upper())
      count = count + 1

      # check if count is equal to group size k, if yes, count reset to zero, the result string add '-'
      if count == k:
        count = 0
        results.append('-')

  if (len(s) > 0 and results[-1] == '-'):  # check if the last character is '-'
    results = results[:-1] # if it is, remove it
  results = results[::-1]   # reverse the whole list
  results = "".join(results) # make a whole list into a string using .join() function
  return results

            

In [ ]:
licenseKeyFormatting("5F3Z-2e-9-w", 4)

In [ ]:
# This should return TRUE
licenseKeyFormatting("5F3Z-2e-9-w", 4) == "5F3Z-2E9W"

In [ ]:
# This should return TRUE
licenseKeyFormatting("2-5g-3-J", 2) == "2-5G-3J"

### [Problem 383](https://leetcode.com/problems/ransom-note/) - Ransom Note

Given two strings `ransomNote` and `magazine`, return `true` if `ransomNote` can be constructed by using the letters from `magazine` and `false` otherwise.

Each letter in `magazine` can only be used once in `ransomNote`.

**Example 1:**
```
Input: ransomNote = "a", magazine = "b"
Output: false
```

**Example 2:**
```
Input: ransomNote = "aa", magazine = "ab"
Output: false
```

**Example 3:**
```
Input: ransomNote = "aa", magazine = "aab"
Output: true
```

**Constraints:**
```
1 <= ransomNote.length, magazine.length <= 105
ransomNote and magazine consist of lowercase English letters.
```

In [ ]:
def count_occurrences(s):
  """compute the occurrence of each letter in the given string, and returns a dictionary with
  letter as key, occurrences as value"""
  counts = {}
  for letter in s:
    if letter in counts:
      # if this letter has already been in the dictionary, add 1 to count 
      counts[letter] = counts[letter] + 1
    else:
      # if not present already, create an entry
      counts[letter] = 1
  return counts

In [ ]:
count_occurrences('aab')

In [ ]:
def canConstruct(s1,s2):
  # create an occurrences dictionary for each string
  d1 = count_occurrences(s1)
  print("d1", d1)
  d2 = count_occurrences(s2)
  print("d2", d2)

  flag = 0
  for letter in d1:
    for character in d2:
      if letter == character and d1[letter] <= d2[character]:
         flag += 1
  if flag == len(d1):
    return True
  else:
    return False


In [ ]:
canConstruct("aabcde","aabccd")

In [ ]:
def canConstruct(ransomNote, magazine):
    def count_occurrences(s):
        counts = {}
        for letter in s:
            if letter in counts:
                # if this letter has already been in the dictionary, add 1 to count 
                counts[letter] = counts[letter] + 1
            else:
                # if not present already, create an entry
                counts[letter] = 1
        return counts
    
    # create an occurrences dictionary for each string
    d1 = count_occurrences(ransomNote)
    d2 = count_occurrences(magazine)

    flag = 0
    for letter in d1:
        for character in d2:
            if letter == character and d1[letter] <= d2[character]:
                flag += 1
    if flag == len(d1):    
        return True
    else:
        return False
    

In [ ]:
# This should return "True"
canConstruct("a", "b") == False

In [ ]:
# This should return "False"
canConstruct("aa", "ab") == True

In [ ]:
# This should return "True"
canConstruct("aa", "aab") == True